dkeyserver/Enclave/enclave.cpp

```diff
@@ -387,7 +335,368 @@ int handle_communication_until_done(
+
+            log_d(TLS_SERVER "domainkey received succeed:\n");
+            for (unsigned long int i = 0; i < sizeof(tmp_dk); i++)
+            {
+                log_d(TLS_SERVER "domain_key[%u]=%2u\n", i, tmp_dk[i]);
+            }
+            break;
+        }
+    } while (1);
+done:
+    return ret;
+}
+
+sgx_status_t get_domainkey_from_target(const char *target_server_name, uint16_t target_server_port)
+{
+    log_d(TLS_SERVER "Connecting target server...\n");
+
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    SSL_CTX *ssl_client_ctx = nullptr;
+    SSL *ssl_session = nullptr;
+
+    X509 *cert = nullptr;
+    EVP_PKEY *pkey = nullptr;
+    SSL_CONF_CTX *ssl_confctx = SSL_CONF_CTX_new();
+
+    int client_socket = -1;
+    int error = 0;
+    if (target_server_name == NULL)
+    {
+        log_d(TLS_SERVER "Connecting target failed: target server name unavailable.\n");
+        goto done;
+    }
+
+    if ((ssl_client_ctx = SSL_CTX_new(TLS_client_method())) == nullptr)
+    {
+        log_d(TLS_SERVER "Unable to create a new SSL context\n");
+        goto done;
+    }
+
+    if (initalize_ssl_context(ssl_confctx, ssl_client_ctx) != SGX_SUCCESS)
+    {
+        log_d(TLS_SERVER "Unable to create a initialize SSL context\n ");
+        goto done;
+    }
+
+    // specify the verify_callback for verification
+    SSL_CTX_set_verify(ssl_client_ctx, SSL_VERIFY_PEER, &verify_callback);
+    log_d(TLS_SERVER "Load cert and key\n");
+    if (load_tls_certificates_and_keys(ssl_client_ctx, cert, pkey) != 0)
+    {
+        log_d(TLS_SERVER "Unable to load certificate and private key on the client\n");
+        goto done;
+    }
+
+    if ((ssl_session = SSL_new(ssl_client_ctx)) == nullptr)
+    {
+        log_d(TLS_SERVER "Unable to create a new SSL connection state object\n");
+        goto done;
+    }
+
+    log_d(TLS_SERVER "New ssl connection getting created\n");
+    client_socket = create_socket(target_server_name, target_server_port);
+    if (client_socket == -1)
+    {
+        log_d(
+            TLS_SERVER "Create a socket and initiate a TCP connect to target server: %s:%d "
+            "(errno=%d)\n",
+            target_server_name,
+            target_server_port,
+            errno);
+        goto done;
+    }
+
+    // set up ssl socket and initiate TLS connection with TLS target server
+    if (SSL_set_fd(ssl_session, client_socket) != 1)
+    {
+        log_d(TLS_SERVER "Ssl set fd error.\n");
+	goto done;
+    }
+
+    if ((error = SSL_connect(ssl_session)) != 1)
+    {
+        log_d(TLS_SERVER "Error: Could not establish a TLS session ret2=%d "
+                       "SSL_get_error()=%d\n",
+            error,
+            SSL_get_error(ssl_session, error));
+        goto done;
+    }
+    log_d(TLS_SERVER "successfully established TLS channel:%s\n",
+        SSL_get_version(ssl_session));
+
+    // start the communication
+    if ((error = get_dkey_from_target(ssl_session)) != 0)
```

Comment:
seems this name is duplicate with the caller: get_domainkey_from_target

do we need to split the logic into a separate function? i think they are doing the same thing?

Reply:
done.